yt_dlp/extractor/adn.py

```diff
@@ -176,16 +176,29 @@ def _perform_login(self, username, password):
 
     def _real_extract(self, url):
         lang, video_id = self._match_valid_url(url).group('lang', 'id')
+        username, password = self._get_login_info()
+        # ADN requires login for German site (at least for now)
+        if lang == 'de' and (not username or not password):
+            self.raise_login_required(method='password')
+
         video_base_url = self._PLAYER_BASE_URL + 'video/%s/' % video_id
```

Comment:
since you said this in the linked issue:

> Currently there is no free German content, so we could just hardcode the login requirement for .de Links. But this may change once the site launches officially. There was free content a few weeks ago, so this may change at any time.

I don't think we should hardcode this behavior

We should check for the truthiness of `formats` and `self._HEADERS` after format extraction, and `raise_login_required` if both are falsy
```suggestion
        video_base_url = self._PLAYER_BASE_URL + 'video/%s/' % video_id
```

Reply:
Makes sense, only case that won't be handled like this is when you have an account but no subscription. I'm going to asume that there is only this case for `formats` being falsy, so I'd suggest this:
```python
        if not formats:
            if not self._HEADERS:
                self.raise_login_required(method='password')

            raise ExtractorError(subscription_msg, expected=True)
```

Another thing I've noticed while playing around with different cases: I'll also add another check for not yet released videos, as that will currently raise a 'subscription required' error.

Reply:
Or even simpler, since you need a subscription even when you are loggend in:
```python
        if not formats:
            self.raise_login_required(msg='This video requires a subscription', method='password')
```
And the same also above